lib/filewatch/since_db.rb

```diff
@@ -0,0 +1,247 @@
+        old_key, new_key = wf.first_fingerprint_match_any?(vk.sort)
+        # old_key is the one found via short_keys
+        # new_key is the altered wf.storage key
+        # it would have a new shorter fingerprint
+        break if old_key && new_key # can't set during iteration
+      end
+      if old_key && new_key && member?(old_key)
+        @logger.debug? && @logger.debug("SinceDb find: short keys match - #{old_key}, #{new_key}")
+        value = delete(old_key)
+        if value.watched_file.nil?
+          @logger.debug? && @logger.debug("SinceDb find: short keys unallocated wf")
+          wf.update_bytes_read(value.position)
+          value.upd_watched_file(wf)
+        elsif value.watched_file != wf
+          @logger.debug? && @logger.debug("SinceDb find: short keys allocated wf unequal")
+          # put back
+          set(old_key, value)
+          # new value
+          value = SincedbValue.new(0) # allow the new allocation to start from scratch
+          value.upd_watched_file(wf)
+        else
+          @logger.debug? && @logger.debug("SinceDb find: short keys allocated wf equal")
+          # its the same watched file with a new key
+          # nothing to do
+        end
+        set(new_key, value)
+        @logger.debug? && @logger.debug("SinceDb find: found! - #{value}")
+        return value
+      end
+      @logger.debug? && @logger.debug("SinceDb find: NOT FOUND!, path: #{wf.path}")
+      nil
+    end
+
+    def member?(key)
+      @sincedb.member?(key)
+    end
+
+    def move(k1, k2)
+      set(k2, delete(k1))
+    end
+
+    def get(key)
+      @sincedb[key]
+    end
+
+    def delete(key)
+      @sincedb.delete(key)
+    end
+
+    def last_read(key)
+      @sincedb[key].position
+    end
+
+    def rewind(key)
+      @sincedb[key].upd_position(0)
+    end
+
+    def store_last_read(key, last_read)
+      @sincedb[key].upd_position(last_read)
+    end
+
+    def increment(key, amount)
+      @sincedb[key].inc_position(amount)
+    end
+
+    def add_watched_file(key, wf)
+      @sincedb[key].upd_watched_file(wf)
+    end
+
+    def clear
+      @sincedb.clear
+    end
+
+    def keys
+      @sincedb.keys
+    end
+
+    def set(key, value)
+      @sincedb[key] = value
+      if key.version?(2) && key.short?
+        @short_keys[key.size] << key
+      end
+      value
+    end
+
+    def unallocated?(key)
+      return false if !member?(key)
+      get(key).watched_file.nil?
+    end
+
+    private
+
+    def store_from_file(line)
+      key, value = self.class.parse_line(line)
+      if @time_sdb_opened < value.expires
+        @logger.debug? && @logger.debug("SinceDb open: setting #{key.inspect} to #{value.inspect}")
+        set(key, value)
+      else
+        @logger.debug? && @logger.debug("SinceDb open: expired, ignoring #{key.inspect} to #{value.inspect}")
+      end
+    end
+
+    def serialize
+      @sincedb.map do |key, value|
+        self.class.serialize_kv(key, value)
```

Comment:
as this function isn't used anywhere else, I'd rather have `@sincedb.map { |key, value| "#{key} #{value}" }.join("\n") + "\n"` instead, but no strong feelings here.